auto-cite/auto-cite.py

```diff
@@ -38,11 +39,15 @@
             log(message, 3, "red")
             exit(1)
 
+        # run plugin
         plugin_sources = import_module(f"plugins.{name}").main(data)
 
         log(f"Got {len(plugin_sources)} sources", 2, "green")
 
         for source in plugin_sources:
+            # make unique key for cache matching
+            source["_cache"] = sha256({**source, "plugin": name, "input": file})
+            # add source
```

Comment:
Is there a reason to implement your own cache rather than using `functools.cache` (or `joblib.Memory` if you want a persistent on-disk cache)? Your use case is a little bit different from what I normally use caches for, so there probably is a reason. I just personally try to avoid writing my own caches because I don't want to work out how to handle hash collisions.

Reply:
Though to be fair, python doesn't like to hash dicts, so you'd have to use something like `json.dumps(d, sort_keys=True)` to get a hashable string and use that as your input to the hypothetical cache-enabled `get_citation` function

Reply:
Will follow up on this in a few days.

Reply:
I created #123 to experiment with and discuss this.